picard/util/pipe.py

```diff
@@ -0,0 +1,206 @@
+# -*- coding: utf-8 -*-
+#
+# Picard, the next-generation MusicBrainz tagger
+#
+# Copyright (C) 2022 skelly37
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+
+
+import concurrent.futures
+import os
+from sys import platform as PLATFORM
+
+
+if PLATFORM == "win32" or PLATFORM == "cygwin":
+    import win32pipe  # type: ignore
+    import win32file  # type: ignore
+    from pywintypes import error as WinApiError  # type: ignore
+
+    PLATFORM = "windows"
+
+
+class Pipe:
+    NO_RESPONSE_MESSAGE: str = "No response from FIFO"
+    NOT_FOUND_MESSAGE: str = "FIFO doesn't exist"
+    MESSAGE_TO_IGNORE: str = "Ignore this message, just testing the pipe"
+
+    def __init__(self, app_name: str, app_version: str, args=[]):
+        self.__app_name: str = app_name
+        self.__app_version: str = app_version
+        self.__platform: str = PLATFORM
+
+        self.path: str = self.__generate_filename()
+
+        self.is_pipe_owner: bool = False
+
+        # test if pipe is listened to even if no args provided
+        if type(args) == list:
+            if len(args) == 0:
+                args.append(self.MESSAGE_TO_IGNORE)
+        else:
+            raise ValueError("args argument MUST be a list")
+
+        if self.__platform == "windows":
+            for arg in args:
+                if not self.send_to_pipe(arg):
+                    self.is_pipe_owner = True
+                    break
+        else:
+            if self.__pipe_exists():
+                for arg in args:
+                    if not self.send_to_pipe(arg):
+                        os.unlink(self.path)
```

Comment:
`os.unlink()` may fail, we should handle exceptions

Reply:
Do you have any ideas of exceptions that shouldn't stop the program execution or is the current commit ok?

Reply:
@phw do you have any ideas other than ones already included in the file or can we mark this as resolved?

Reply:
It currently handles `FileNotFoundError`, so that's fine.

I'm not sure what to do in other cases, like e.g. `PermissionErrors`. There are two general approaches we can go if something with pipe creation fails seriously in a non-fixable / unexpected manner:

- we crash the application
- we log an error, but we start Picard still as a separate process without pipe integration

I tend to prefer option 2 for this case, because it allows user to still run Picard, they will just loose the single instance advantages.

Reply:
> I tend to prefer option 2 for this case

Me too.

Reply:
So my idea is to:
1. create sth like `self.permission_error_happened` in except block
2. If the previously mentioned attribute is true, pass it to the Tagger's constructor the same way as if the `--force` flag was used.

Reply:
> So my idea is to:
> 
>     1. create sth like `self.permission_error_happened` in except block
> 
>     2. If the previously mentioned attribute is true, pass it to the Tagger's constructor the same way as if the `--force` flag was used.

Done